Added one more chapter to starter guide (#2238)
zenml-io/zenml: docs/book/user-guide/starter-guide/structuring-a-project.md

You will learn more about different environments in a [later chapter](../advanced-guide/environment-management/environment-management.md).
{% endhint %}

## Pattern 2: Artifact exchange between pipelines through a `Model`

While passing around artifacts with ID's or names is very useful, often times it is desirable to use the `ZenML Model` be the point of reference instead. 

For example, let's say we have a training pipeline called `train_and_promote` and an inference pipeline called `do_predictions`. The training pipeline produces many different model artifacts, all of which are collected within a [ZenML Model](track-ml-models.md). Each time the `train_and_promote` pipeline runs, it creates a new `iris_classifier`. However, it only promotes the Model Version to `production` if a certain accuracy threshold is met. The promotion can be also be done manually with human intervention

> On the other side, the `do_predictions` pipeline simply picks up the latest promoted model and runs batch inference on it. It need not know of the ID's or names of any of the artifacts produced by the training pipeline's many runs. This way these two pipelines can independently be run, but can rely on each other's output.

Reviewer: ```suggestion
On the other side, the `do_predictions` pipeline simply picks up the latest promoted model and runs batch inference on it. It need not know of the IDs or names of any of the artifacts produced by the training pipeline's many runs. This way these two pipelines can independently be run, but can rely on each other's output.
```

```python
from typing_extensions import Annotated
from zenml import get_pipeline_context, pipeline, ExternalArtifact
from zenml.enums import ModelStages
from zenml.model import ModelVersion
import pandas as pd
from sklearn.base import ClassifierMixin

@step
def predict(
    model: ClassifierMixin,
    data: pd.DataFrame,
) -> Annotated[pd.Series, "predictions"]:
    predictions = pd.Series(model.predict(data))
    return predictions

@pipeline(
    model_config=ModelVersion(
        name="iris_classifier",
        # Using the production stage
        version=ModelStages.PRODUCTION,
    ),
)
def do_predictions():
    # model_name and model_version derived from pipeline context
    model_version = get_pipeline_context().model_version
    inference_data = load_data()
    predict(
        # Here, we load in the `trained_model` from a trainer step
        model=model_version.get_model_artifact("trained_model"),  
        data=inference_data,
    )

if __name__ == "__main__":
    do_predictions()
```

> Ultimately, both approaches are fine. Users should decide which one to use based on their own preferences.

Reviewer: ```suggestion
Ultimately, both approaches are fine. You should decide which one to use based on your own preferences.
```
